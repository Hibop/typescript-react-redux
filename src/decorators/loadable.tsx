import * as React from 'react'

export default function loadable<P>(isLoading: (p: P) => boolean) {
  return (C: React.ComponentClass<P>|React.SFC<P>) => {
    const LoadableComponent: React.SFC<P> = (props) => {
      if (isLoading(props)) {
        return <div>Just a moment, please...</div>
      }
      return <C {...props} />
    }

    // Set pretty `displayName` for dev tooling
    LoadableComponent.displayName = `Loadable(${C.name})`
    return LoadableComponent
  }
}
