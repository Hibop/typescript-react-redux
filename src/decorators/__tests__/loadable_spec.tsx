// tslint:disable-next-line no-unused-variable
import * as React from 'react'
import * as TestUtils from 'react-dom/test-utils'
import * as ReactShallowRenderer from 'react-test-renderer/shallow'

import loadable from '../loadable'

const TestComponent = () => <div>OK</div>
const LoadableTestComponent = loadable(props => props.isLoading)(TestComponent)

describe('decorators/loadable', () => {

  it('sets friendly .displayName', () => {
    expect(LoadableTestComponent.displayName)
      .toEqual('Loadable(TestComponent)')
  })

  it('maps props to loading', () => {
    const renderer = new ReactShallowRenderer()
    expect(renderer.render(<LoadableTestComponent isLoading />).props.children)
      .toEqual('Just a moment, please...')

    expect(renderer.render(<LoadableTestComponent />))
      .toMatchSnapshot()
  })
})
