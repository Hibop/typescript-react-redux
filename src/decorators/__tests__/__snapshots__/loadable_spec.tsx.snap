// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorators/loadable maps props to loading 1`] = `<TestComponent />`;
